lib/maths/CSignal.cc

```diff
@@ -182,12 +180,12 @@ void CSignal::autocorrelations(const TFloatMeanAccumulatorVec& values, TDoubleVe
     double mean = CBasicStatistics::mean(moments);
     double variance = CBasicStatistics::maximumLikelihoodVariance(moments);
 
-    TComplexVec f;
-    f.reserve(n);
+    TComplexVec f(n, TComplex{});
```

Comment:
Yeah, this change was just an attempt to shut `Valgrind` up about the following...
```
Conditional jump or move depends on uninitialised value(s)
std::__1::complex<double> std::__1::operator*<double>(std::__1::complex<double> const&, std::__1::complex<double> const&)
std::__1::complex<double>& std::__1::complex<double>::operator*=<double>(std::__1::complex<double> const&)
ml::maths::CSignal::hadamard(std::__1::vector<std::__1::complex, std::__1::allocator> const&, std::__1::vector<std::__1::complex, std::__1::allocator>&)
ml::maths::CSignal::fft(std::__1::vector<std::__1::complex, std::__1::allocator>&)
ml::maths::CSignal::autocorrelations(std::__1::vector<ml::maths::CBasicStatistics::SSampleCentralMoments, std::__1::allocator> const&, std::__1::vector<double, std::__1::allocator>&)
ml::maths::(anonymous namespace)::mostSignificantPeriodicComponent(std::__1::vector<ml::maths::CBasicStatistics::SSampleCentralMoments, std::__1::allocator>)
ml::maths::testForPeriods(ml::maths::CPeriodicityHypothesisTestsConfig const&, long, long, std::__1::vector<ml::maths::CBasicStatistics::SSampleCentralMoments, std::__1::allocator> const&)
ml::maths::CTimeSeriesDecompositionDetail::CPeriodicityTest::test(ml::maths::CTimeSeriesDecompositionDetail::SAddValue const&)
ml::maths::CTimeSeriesDecompositionDetail::CPeriodicityTest::handle(ml::maths::CTimeSeriesDecompositionDetail::SAddValue const&)
ml::maths::CTimeSeriesDecomposition::addPoint(long, double, std::__1::array<double, 4ul> const&, std::__1::function<void (std::__1::vector)> const&)
CTimeSeriesDecompositionTest::testMixedSmoothAndSpikeyDataProblemCase::test_method()
CTimeSeriesDecompositionTest::testMixedSmoothAndSpikeyDataProblemCase_invoker()
Uninitialised value was created by a heap allocation
malloc
operator new(unsigned long)
std::__1::__libcpp_allocate(unsigned long, unsigned long)
std::__1::allocator<std::__1::complex>::allocate(unsigned long, void const*)
std::__1::allocator_traits<std::__1::allocator>::allocate(std::__1::allocator<std::__1::complex>&, unsigned long)
std::__1::__split_buffer<std::__1::complex, std::__1::allocator&>::__split_buffer(unsigned long, unsigned long, std::__1::allocator<std::__1::complex>&)
std::__1::__split_buffer<std::__1::complex, std::__1::allocator&>::__split_buffer(unsigned long, unsigned long, std::__1::allocator<std::__1::complex>&)
std::__1::vector<std::__1::complex, std::__1::allocator>::reserve(unsigned long)
ml::maths::CSignal::autocorrelations(std::__1::vector<ml::maths::CBasicStatistics::SSampleCentralMoments, std::__1::allocator> const&, std::__1::vector<double, std::__1::allocator>&)
ml::maths::(anonymous namespace)::mostSignificantPeriodicComponent(std::__1::vector<ml::maths::CBasicStatistics::SSampleCentralMoments, std::__1::allocator>)
ml::maths::testForPeriods(ml::maths::CPeriodicityHypothesisTestsConfig const&, long, long, std::__1::vector<ml::maths::CBasicStatistics::SSampleCentralMoments, std::__1::allocator> const&)
ml::maths::CTimeSeriesDecompositionDetail::CPeriodicityTest::test(ml::maths::CTimeSeriesDecompositionDetail::SAddValue const&)
```

Maybe `Valgrind` just got confused? If so, as I said in a comment above, I can always just add a suppression for this warning.

Reply:
Let me study the code a bit and have a think about this. I can't at present see any reason why the loop should be triggering an error so may be a false positive, but I think we would need an alternative fix if we wanted to tackle this. One possibility would be to presize then not extend, but just write values into place.

Reply:
> Let me study the code a bit and have a think about this. I can't at present see any reason why the loop should be triggering an error so may be a false positive, but I think we would need an alternative fix if we wanted to tackle this. One possibility would be to presize then not extend, but just write values into place.

Thanks Tom. Welcome back btw :-)

Reply:
Along these lines, how about the following:
```
TComplexVec f{n, TComplex{0.0, 0.0}};
for (std::size_t i = 0; i < n; ++i) {
    std::size_t j = i;
    for (/**/; j < n && CBasicStatistics::count(values[j]) == 0; ++j) {
        // no-op
    }
    if (i < j) {
        // Infer missing values by linearly interpolating.
        if (j == n) {
            break;
        } else if (i > 0) {
            for (std::size_t k = i; k < j; ++k) {
                double alpha{static_cast<double>(k - i + 1) / static_cast<double>(j - i + 1)};
                double real{CBasicStatistics::mean(values[j]) - mean};
                f[k] = (1.0 - alpha) * f[i - 1] + alpha * TComplex{real, 0.0};
            }
        }
        i = j;
    }
    f[i] = TComplex{CBasicStatistics::mean(values[i]) - mean, 0.0};
}
```
This is actually clearer IMO and may incidentally fix the valgrind error.

Reply:
Thanks Tom, giving that a go now!

Reply:
Also, do you think:

```
    while (j < n && CBasicStatistics::count(values[j]) == 0) {
        ++j;
    }
```

is clearer than:

```
    for (/**/; j < n && CBasicStatistics::count(values[j]) == 0; ++j) {
        // no-op
    }
```

Reply:
I do :-) Making the change...

Reply:
> Yeah, this change was just an attempt to shut `Valgrind` up about the following...
>```
>Conditional jump or move depends on uninitialised value(s)
>std::__1::complex<double> std::__1::operator*<double>(std::__1::complex<double> const&, >std::__1::complex<double> const&)
>std::__1::complex<double>& std::__1::complex<double>::operator*=<double>(std::__1::complex<double> const&)
>ml::maths::CSignal::hadamard(std::__1::vector<std::__1::complex, std::__1::allocator> const&, std::__1::vector<std::__1::complex, std::__1::allocator>&)
>```

It's worth noting that `hadamard` would produce an uninitialized memory read if either:

1. `fx` is bigger than `fy`
2. `fx` or `fy` contains uninitialized memory

The inputs to `hadamard` (`fx` and `fy`) might have been produced by `radix2fft` (because `fft` calls it in some cases), so the error in `radixfft` could have been causing the Valgrind error in `hadamard`.

Did you see the error above before or after the error in `radix2fft` was fixed?

Reply:
That also looked odd to me. The offending call was from `fft`, but it is applied to `a` and `b` which are initialised as
```
TComplexVec a(m, TComplex(0.0));
TComplexVec b(m, TComplex(0.0));
``` 
so must be of the same size. The only other thing I thought on this was does calling `TComplex(0.0)` leave the imaginary part uninitialised, I haven't checked? May be clearer anyway to change to `TComplex{0.0, 0.0}` in these definitions.

Reply:
> Did you see the error above before or after the error in radix2fft was fixed?

@droberts195 I _think_ the error was visible after the `radix2fft` was fixed (at least I didn't notice it in the initial Valgrind report)

Reply:
Also @tveasey, your suggested change above ^^^ looks good. No Valgrind errors in `CTimeSeriesDecompositionTest/testMixedSmoothAndSpikeyDataProblemCase` at least.

Reply:
> so must be of the same size

`hadamard` is a public method, so not necessarily called from there.  Also, a bug in another method called in between initialization and the call to `hadamard` could accidentally change the size of one of the vectors.  Saying "so must be of the same size" assumes the reader is intimately familiar with all possible code paths.

> does calling `TComplex(0.0)` leave the imaginary part uninitialised

No, it initializes it to `0.0`.  I checked the library code when I first opened #845.  But I agree it's clearer and equally efficient to explicitly pass both arguments.

Reply:
> hadamard is a public method, so not necessarily called from there. 

But the stack trace includes `CSignal::fft`